File: artifacts/tulipindicators/decompose/src/md.c
int ti_md_start(const double *options)
{
  return ((int) options[0]) - 1;
}


/*** DEPENDENCIES:
***/


int ti_md(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *input = inputs[0];
  const int period = (int) options[0];
  double *output = outputs[0];
  unsigned int output_idx = 0;
  const double scale = 1.0 / period;
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_md_start(options))
  {
    return 0;
  }
  double sum = 0;
  int i;
  int j;
  for (i = 0; i < size; i += 1)
  {
    const double today = input[i];
    sum += today;
    if (i >= period)
    {
      sum -= input[i - period];
    }
    const double avg = sum * scale;
    if (i >= (period - 1))
    {
      double acc = 0;
      for (j = 0; j < period; j += 1)
      {
        acc += fabs(avg - input[i - j]);
      }

      output[output_idx] = acc * scale;
      output_idx += 1;
    }
  }

  assert(((&output[output_idx]) - outputs[0]) == (size - ti_md_start(options)));
  return 0;
}


/*** DEPENDENCIES:
int ti_md_start(const double *options)
{
  return ((int) options[0]) - 1;
}


----------------------------
***/

File: artifacts/tulipindicators/translate/src/md.rs
use crate::*;
use std::f64;

pub fn ti_md_start(options: Option<&[f64]>) -> Option<i32> {
    // Check if options is None (equivalent to NULL pointer check in C)
    let options = options?;
    
    // Ensure the slice has at least one element to avoid panic
    if options.is_empty() {
        return None;
    }

    // Perform the same calculation as in C
    Some(options[0] as i32 - 1)
}

pub fn ti_md(
    size: i32,
    inputs: Option<&[&[f64]]>,
    options: Option<&[f64]>,
    outputs: Option<&mut [&mut [f64]]>,
) -> Option<i32> {
    // Check for None in all Option parameters
    let inputs = inputs?;
    let options = options?;
    let outputs = outputs?;

    // Get slices from the input arrays
    let input = inputs[0];
    let period = options[0] as i32;
    let output = &mut outputs[0];

    if period < 1 {
        return Some(1);
    }

    let md_start = ti_md_start(Some(options))?;
    if size <= md_start {
        return Some(0);
    }

    let scale = 1.0 / period as f64;
    let mut sum = 0.0;
    let mut output_idx = 0;

    for i in 0..size as usize {
        let today = input[i];
        sum += today;

        if i >= period as usize {
            sum -= input[i - period as usize];
        }

        let avg = sum * scale;

        if i >= (period - 1) as usize {
            let mut acc = 0.0;
            for j in 0..period as usize {
                acc += f64::abs(avg - input[i - j]);
            }

            output[output_idx] = acc * scale;
            output_idx += 1;
        }
    }

    // Note: The assert in the original C code is omitted as it's a debug check
    Some(0)
}
